src/core/auth/qgsauthmethodregistry.cpp

```diff
@@ -0,0 +1,392 @@
+    output->setTitle( QObject::tr( "No Authentication Methods" ) );
+    output->setMessage( msg, QgsMessageOutput::MessageText );
+    output->showMessage();
+    return;
+  }
+
+  // auth method file regex pattern, only files matching the pattern are loaded if the variable is defined
+  QString filePattern = getenv( "QGIS_AUTHMETHOD_FILE" );
+  QRegExp fileRegexp;
+  if ( !filePattern.isEmpty() )
+  {
+    fileRegexp.setPattern( filePattern );
+  }
+
+  QListIterator<QFileInfo> it( mLibraryDirectory.entryInfoList() );
+  while ( it.hasNext() )
+  {
+    QFileInfo fi( it.next() );
+
+    if ( !fileRegexp.isEmpty() )
+    {
+      if ( fileRegexp.indexIn( fi.fileName() ) == -1 )
+      {
+        QgsDebugMsg( "auth method " + fi.fileName() + " skipped because doesn't match pattern " + filePattern );
+        continue;
+      }
+    }
+
+    QLibrary myLib( fi.filePath() );
+    if ( !myLib.load() )
+    {
+      QgsDebugMsg( QString( "Checking %1: ...invalid (lib not loadable): %2" ).arg( myLib.fileName() ).arg( myLib.errorString() ) );
+      continue;
+    }
+
+    // get the description and the key for the auth method plugin
+    isauthmethod_t *isAuthMethod = ( isauthmethod_t * ) cast_to_fptr( myLib.resolve( "isAuthMethod" ) );
+    if ( !isAuthMethod )
+    {
+      QgsDebugMsg( QString( "Checking %1: ...invalid (no isAuthMethod method)" ).arg( myLib.fileName() ) );
+      continue;
+    }
+
+    // check to see if this is an auth method plugin
+    if ( !isAuthMethod() )
+    {
+      QgsDebugMsg( QString( "Checking %1: ...invalid (not an auth method)" ).arg( myLib.fileName() ) );
+      continue;
+    }
+
+    // looks like an auth method plugin. get the key and description
+    description_t *pDesc = ( description_t * ) cast_to_fptr( myLib.resolve( "description" ) );
+    if ( !pDesc )
+    {
+      QgsDebugMsg( QString( "Checking %1: ...invalid (no description method)" ).arg( myLib.fileName() ) );
+      continue;
+    }
+
+    methodkey_t *pKey = ( methodkey_t * ) cast_to_fptr( myLib.resolve( "authMethodKey" ) );
+    if ( !pKey )
+    {
+      QgsDebugMsg( QString( "Checking %1: ...invalid (no authMethodKey method)" ).arg( myLib.fileName() ) );
+      continue;
+    }
+
+    // add this auth method to the method map
+    mAuthMethods[pKey()] = new QgsAuthMethodMetadata( pKey(), pDesc(), myLib.fileName() );
+
+  }
+}
+
+// typedef for the unload auth method function
+typedef void cleanupAuthMethod_t();
+
+QgsAuthMethodRegistry::~QgsAuthMethodRegistry()
+{
+  AuthMethods::const_iterator it = mAuthMethods.begin();
+
+  while ( it != mAuthMethods.end() )
+  {
+    QgsDebugMsg( QString( "cleanup: %1" ).arg( it->first ) );
+    QString lib = it->second->library();
+    QLibrary myLib( lib );
+    if ( myLib.isLoaded() )
+    {
+      cleanupAuthMethod_t* cleanupFunc = ( cleanupAuthMethod_t* ) cast_to_fptr( myLib.resolve( "cleanupAuthMethod" ) );
+      if ( cleanupFunc )
+        cleanupFunc();
+    }
```

Comment:
i think this leaks the metadata created in line 140 - needs a 

    delete it->second;